teps/0130-pipeline-level-build-cache.md

```diff
@@ -0,0 +1,715 @@
+---
+status: proposed
+title: Pipeline-level Build Cache
+creation-date: '2023-01-24'
+last-updated: '2023-02-21'
+authors:
+- '@lbernick'
+collaborators: []
+---
+
+# TEP-0130: Pipeline-level Build Cache
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Out of Scope](#out-of-scope)
+  - [Notes and Caveats](#notes-and-caveats)
+    - [Other uses for this feature](#other-uses-for-this-feature)
+    - [Related features in other CI/CD systems](#related-features-in-other-cicd-systems)
+- [Design Details](#design-details)
+- [Design Evaluation](#design-evaluation)
+  - [Reusability](#reusability)
+  - [Simplicity](#simplicity)
+  - [Flexibility](#flexibility)
+  - [Conformance](#conformance)
+  - [User Experience](#user-experience)
+  - [Performance](#performance)
+  - [Security](#security)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+  - [Improve our documentation](#improve-our-documentation)
+  - [Catalog Tasks for docker daemon](#catalog-tasks-for-docker-daemon)
+  - [First-class docker support](#first-class-docker-support)
+- [Implementation Plan](#implementation-plan)
+  - [Test Plan](#test-plan)
+  - [Infrastructure Needed](#infrastructure-needed)
+  - [Upgrade and Migration Strategy](#upgrade-and-migration-strategy)
+  - [Implementation Pull Requests](#implementation-pull-requests)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+This TEP proposes support for more complex build Pipelines by allowing multiple Tasks in
+a PipelineRun to share the same build cache; for example, using the same docker daemon.
+
+## Motivation
+
+The docker-build catalog Task only allows building a single image and pushing it to a registry.
+The following use cases are not addressed:
+- Building multiple images concurrently with the same daemon to save resources and share cached base layers
```

Comment:
What use case is driving this proposal? What is the proposed storage for such images?

Reply:
The use cases are listed under the motivation section:
- User wants to test an image after building, and only push images that pass tests to the artifact repo
- User wants to save resources by using the same docker daemon/cached layers for multiple images built in the same pipelinerun (e.g. via matrix)

Image layers would be cached by the docker daemon (as it does currently). The user is still responsible for pushing the built image to a repository.

Reply:
thanks @lbernick yup I understand the two bullet points listed in the TEP as well. My question was more aligning with @vdemeester's feedback/comment, should we propose a feature in the pipeline controller for one single technology - docker daemon?

Reply:
I'm hoping that the reframing of this TEP as "pipeline level service" makes it generic enough. Yes, docker daemon is just one tool, but docker builds are a very common CI/CD use case and I think it's important to consider features that will improve the UX for docker builds. I also listed a few more potential uses for a Pipeline level service, although I think they're less compelling.